### DISCUSSION TITLE
How to show element only once but any of them?

### DISCUSSION THREAD

**Discussion by Serhioromano** [eclipse-langium/langium#790] [Q&A] [answered]
I have code that can be

```
VAR_GLOBAL
END_VAR_GLOBAL

VAR_GLOBAL RETAIN
END_VAR_GLOBAL

VAR_GLOBAL CONSTANT
END_VAR_GLOBAL

VAR_GLOBAL CONSTANT PERSISTENT
END_VAR_GLOBAL
```

So it might have one or number for keywords but every keywork only once, here is what I do.

```
VarGlobal:
    'VAR_GLOBAL' (decconstant+=DeclarationConstant)?
    'END_VAR_GLOBAL'
;

DeclarationConstant returns string:
    'CONSTANT' | 'RETAIN' | 'PERSISTENT'
```

But in this case I can only place one keyword, How is it done correctly?

**Comment by msujew** [comment, marked as answer]
Hey @Serhioromano,

so in theory you could use the unordered groups feature:

```
VarGlobal:
    'VAR_GLOBAL' (decconstant+='CONSTANT' & decconstant+='RETAIN' & decconstant+='PERSISTENT')
    'END_VAR_GLOBAL'
;
```

However, I would recommend a more flexible approach with better error messages:

```
VarGlobal:
    'VAR_GLOBAL' (decconstant+=DeclarationConstant)* // Allow an unlimited amount of modifiers
    'END_VAR_GLOBAL'
;

DeclarationConstant returns string:
    'CONSTANT' | 'RETAIN' | 'PERSISTENT'
```

And then simply write a custom validator to determine whether modifiers are duplicated. That allows you to write a specific message that helps users identify issues (i.e. something like `Duplicate declaration constant modifier`). If you want to purely address this feature by using parser features, users would encounter a cryptic parser error instead (i.e. `Expected X but found Y`).

**Reply by Serhioromano** [reply]
Works perfectly. Thank you.
